### PR TITLE
pkg/k8s/watchers: set ResourceVersion in k8s meta GetOptions{}

### PR DESCRIPTION
**Pull request by ArthurChiao**
Set ResourceVersion=0 will make kube-apiserver serve the request with its own in-memory cache, instead of invoking etcd.KV.Get() to retrieve the information from etcd. The former will only invoke an etcd request only if its own in-memory cache hasn't been ready.

This not only makes the request a little faster, but also reduces burden for k8s etcd, as we can skip #pods GET requests for the latter.

Note that for clients initialized using client-go informer, the RV=0 parameter has already been set by the latter internally.

Signed-off-by: ArthurChiao <arthurchiao@hotmail.com>

### CODE REVIEW

**Comment by ArthurChiao** [comment]
Simple tests with and without `resourceVersion=0` in the request:

```shell
# Util
$ cat curl-k8s-apiserver.sh
curl -s --cert /etc/kubernetes/pki/admin.crt --key /etc/kubernetes/pki/admin.key --cacert /etc/kubernetes/pki/ca.crt $@

# Test 1
for n in {1..1000}; do echo $n; ./curl-k8s-apiserver.sh "https://localhost:6443/apis/cilium.io/v2/namespaces/default/ciliumendpoints/smoke-pod" 1 > /dev/null; done

# Test 2
for n in {1..1000}; do echo $n; ./curl-k8s-apiserver.sh "https://localhost:6443/apis/cilium.io/v2/namespaces/default/ciliumendpoints/smoke-pod?resourceVersion=0" 1 > /dev/null; done
```

k8s-etcd requests received (two `test1` followed by one `test2`):

image

We have a detailed blog for how RV works, unfortunately it's only available in Chinese: https://arthurchiao.art/blog/k8s-reliability-list-data-zh/

**Comment by nathanjsweet** [comment]
I'm a little bit nervous about this one. It is very important that the agent and controller have the same version of a lot of objects as there is important synchronization logic between them. This is particularly true for Cilium-managed resources (like CEPs, CNPs, CiliumIdentities, etc).

**Comment by ArthurChiao** [comment]
Hi @nathanjsweet , sorry for late reply. From your comments, may I assume that the current behavior (also retrieve data from k8s etcd instead of kube-apiserver cache) is expected?

My main concern is that during extreme cases like https://github.com/cilium/cilium/pull/21213 and https://github.com/cilium/cilium/pull/20706, where all agents restart at the same time and you have hundreds of thounsands of pods/endpoints in the cluster, this behavior will results in a request storm to k8s etcd, which may stuck or crash the entire k8s control plane.

**Comment by joestringer** [comment]
Can there be multiple resource versions over time for the same endpoint resource? What are the semantics of the k8s API in that sort of case? Is it that the `Get()` with 0 resourceversion just fetches any "latest" version of the resource or are there other nuances?

**Comment by joestringer** [comment]
Not sure whether to mark this as "release-note/misc" or "release-note/minor". If minor, we should include a user-facing release note something like "Improve CiliumEndpoint resource usage in kube-apiserver"

**Comment by ArthurChiao** [comment]
> Can there be multiple resource versions over time for the same endpoint resource? What are the semantics of the k8s API in that sort of case? Is it that the Get() with 0 resourceversion just fetches any "latest" version of the resource or are there other nuances?

Hi @joestringer , I revisited the K8s doc https://kubernetes.io/docs/reference/using-api/api-concepts/#the-resourceversion-parameter ,  it has a table for the `get` semantic with `resourceVersion`:

<html>
<body>


resourceVersion unset | resourceVersion="0" | resourceVersion="{value other than 0}"
-- | -- | --
Most Recent | Any | Not older than


</body>
</html>

Where I think,

* `Most recent` is just passing the request to etcd like the current code does;
* `Any` is serving the request with the cache in kube-apiserver memory;
* `Not older than` can select specific versions.

So setting RV=0 can not guarantee to always get the lastest object, if that matters, this patch should be closed.

**Comment by joestringer** [comment]
:pray: thanks for that context, this helps.

> So setting RV=0 can not guarantee to always get the latest object, if that matters, this patch should be closed.

This seems like the key question to answer. I don't know right now, but perhaps you could investigate how this endpoint synchronizer logic works and whether it may make the wrong decision if there is an outdated resource? Otherwise we may need to request a review from someone with more background in the operator logic.

**Comment by ArthurChiao** [comment]
> This seems like the key question to answer. I don't know right now, but perhaps you could investigate how this endpoint synchronizer logic works and whether it may make the wrong decision if there is an outdated resource?

I'll investigate the synchronizer logic and come back if I have any updates, thanks!

**Comment by tommyp1ckles** [comment]
The endpointsynchronizer logic is a bit hard to follow

Endpointsynchronizer gets started on both endpoint creates/restores. So if we're worrying about a "thundering herd" situation I think the more likely situation would be a mass agent restart resulting in endpoint restores calling `RunK8sCiliumEndpointSync(...)`. This will immediately cause a K8s apiserver GET in the `needsInit` clause code-path.

After the initial Get, a DeepCopy is done against the local derived endpoint state. If those two differ then the difference is reconciled using a PATCH request to API server.

In this case, I believe that not having a guarantee of having the latest CEP object might be ok, as the Status will be reconciled if necessary.

A concern I do have, following these changes: https://github.com/cilium/cilium/pull/21768#. Cilium will be more _strict_ in ensuring that local Endpoint UID matches that of the CEP. This is to prevent race conditions when managing Pods of the same namespace/name. In the case of a mismatch, the endpointSync controller will simply error out of its state iteration and be retried later. The idea being that such a state should be temporary, and that the old CEP will be cleaned up and "released" shortly (i.e. old Pods management is completed, or controller CEP GC).

Going by the k8s docs:

"It is possible for the request to return data at a much older resource version that the client has previously observed, particularly in high availability configurations, due to partitions or stale caches. _Clients that cannot tolerate this should not use this semantic._"

What happens in API server when a resource (i.e. of the same namespace/name) is fully recreated resulting in a new UID. _Could using resourceVersion="0" possibly return a completely old resource?_ Or would recreation of the CRD object force an invalidation of it in cache?

**Comment by ArthurChiao** [comment]
Hi @tommyp1ckles, thanks a lot for your detailed information!

> What happens in API server when a resource (i.e. of the same namespace/name) is fully recreated resulting in a new UID. Could using resourceVersion="0" possibly return a completely old resource? 

I suppose yes, here is a hypothetical scenario:

1. Cilium posts new CR to kube-apiserver (which has the same ns/name as before, such as statefulset pod);
2. kube-apiserver posts the object to etcd;
3. etcd persistents the object;
4. kube-apiserver receives the `create` event then updates it to the local cache;

If kube-apiserver loses connectivity to etcd after step 3 but before step 4, and cilium performs get CEP requests with RV=0 in this time period, it will get the old CEP.

> Or would recreation of the CRD object force an invalidation of it in cache?

Based on the above analysis, I don't think so.
